packages/aws-cdk/lib/util/work-graph.ts

```diff
@@ -0,0 +1,216 @@
+/* eslint-disable no-console */
+import { WorkNode, DeploymentState, StackNode, AssetBuildNode, AssetPublishNode } from './work-graph-types';
+
+export class WorkGraph {
+  public readonly nodes: Record<string, WorkNode>;
+  private readonly readyPool: Array<WorkNode> = [];
+  private readonly lazyDependencies = new Map<string, string[]>();
+  public error?: Error;
+
+  public constructor(nodes: Record<string, WorkNode> = {}) {
+    this.nodes = { ...nodes };
+  }
+
+  public addNodes(...nodes: WorkNode[]) {
+    for (const node of nodes) {
+      const ld = this.lazyDependencies.get(node.id);
+      if (ld) {
+        node.dependencies.push(...ld);
+        this.lazyDependencies.delete(node.id);
+      }
+
+      this.nodes[node.id] = node;
+    }
+  }
+
+  /**
+   * Add a dependency, that may come before or after the nodes involved
+   */
+  public addDependency(fromId: string, toId: string) {
+    const node = this.nodes[fromId];
+    if (node) {
+      node.dependencies.push(toId);
+      return;
+    }
+    let lazyDeps = this.lazyDependencies.get(fromId);
+    if (!lazyDeps) {
+      lazyDeps = [];
+      this.lazyDependencies.set(fromId, lazyDeps);
+    }
+    lazyDeps.push(toId);
+  }
+
+  public node(id: string) {
+    const ret = this.nodes[id];
+    if (!ret) {
+      throw new Error(`No node with id ${id} among ${Object.keys(this.nodes)}`);
+    }
+    return ret;
+  }
+
+  public absorb(graph: WorkGraph) {
```

Comment:
what an ugly word lol. Is there any semantic reason you used this over `merge` or just its a cooler word

Reply:
Yes. In my mind, it's ambiguous whether `merge` will mutate the LHS in-place, or return a new value.

`absorb` is very obviously mutating the object.